contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,439 @@
+`RevisionHistoryLimit`, the controller will delete them, in order with respect 
+to the value mapped to their `.Revisions`, until the number 
+of remaining ControllerRevisions is equal to the `RevisionHistoryLimit`.
+
+This ensures that the number of recorded, non-live revisions is less than or 
+equal to the configured `RevisionHistoryLimit`.
+
+### Version Tracking
+Controllers must track the version of the target Object state that corresponds 
+to their generated Objects. This information is necessary to determine which 
+versions are live, and to track which Objects need to be updated during a 
+target state update or rollback. We propose two methods that controllers may 
+use to track live versions and their association with generated Objects.
+
+1. The most straightforward method is labeling. In this method the generated 
+Objects are labeled with the `.Name` of the ControllerRevision object that 
+corresponds to the version of the target Object state that was used to generate 
+them. As we have taken care to ensure the uniqueness of the `.Names` of the 
+ControllerRevisions, this approach is reasonable.
+    - A revision is considered to be live while any generated Object labeled 
+     with its `.Name` is live.
+    - This method has the benefit of providing visibility, via the label, to 
+    users with respect to the historical providence of a generated Object.
+    - The primary drawback is the lack of support for using garbage collection 
+    to ensure that only non-live version snapshots are collected.
+1. Controllers may also use the `OwnerReferences` field of the 
+ControllerRevision to record all Objects that are generated from target Object 
+state version represented by the ControllerRevision as its owners.
+   - A revision is considered to be live while any generated Object that owns 
+   it is live.
+   - This method allows for the implementation of generic garbage collection.
+   - The primary drawback with this method is that the book keeping is complex, 
+   and deciding if a generated Object corresponds to a particular revision 
+   will require testing each Object for membership in the `OwnerReferences` 
+   of all ControllerRevisions.
+1. Note that, since we are labeling the generated Objects to indicate their
+ providence with respect to the version of the controller's target Object state,
+ we are susceptible to downstream mutations by other controllers changing the 
+ controller's product. The best we can do is guarantee that our product meets 
+ the specification at the time of creation. If a third party mutates the product
+ downstream (as long as it does so in a consistent and intentional way), 
+ we don't want to recall it and make it conform to the original specification. 
+ This would cause the controllers to "fight" indefinitely. 
+   
+At the cost of the complexity of implementing both labeling and ownership, 
+controllers may use a combination of both approaches to mitigate the 
+deficiencies of each.
```

Comment:
We need to the owner reference to ensure that overlap does not occur. We have been bitten in the past by not ensuring the providence of generated objects.